pkg/storage/snapshot/restore.go

```diff
@@ -596,6 +606,115 @@ func (t *vmRestoreTarget) reconcileDataVolumes() (bool, error) {
 	return createdDV || waitingDV, nil
 }
 
+func (t *vmRestoreTarget) getControllerRevision(namespace, name string) (*appsv1.ControllerRevision, error) {
+	revisionKey := cacheKeyFunc(namespace, name)
+	obj, exists, err := t.controller.CRInformer.GetStore().GetByKey(revisionKey)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, fmt.Errorf("Unable to find ControllerRevision %s", revisionKey)
+	}
+	return obj.(*appsv1.ControllerRevision), nil
+}
+
+func (t *vmRestoreTarget) getVirtualMachineSnapshot(namespace, name string) (*snapshotv1.VirtualMachineSnapshot, error) {
+	vmSnapshotKey := cacheKeyFunc(namespace, name)
+	obj, exists, err := t.controller.VMSnapshotInformer.GetStore().GetByKey(vmSnapshotKey)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, fmt.Errorf("Unable to find VirtualMachineSnapshot %s", vmSnapshotKey)
+	}
+	return obj.(*snapshotv1.VirtualMachineSnapshot), nil
+}
+
+func (t *vmRestoreTarget) restoreInstancetypeControllerRevision(revisionName, vmSnapshotName string, vm *kubevirtv1.VirtualMachine) (*appsv1.ControllerRevision, error) {
+	existingCR, err := t.getControllerRevision(vm.Namespace, revisionName)
+	if err != nil {
+		return nil, err
+	}
+
+	vmSnapshot, err := t.getVirtualMachineSnapshot(vm.Namespace, vmSnapshotName)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the CR is controlled by the VirtualMachineSnapshot restore it and attempt to claim it if the VirtualMachine exists
+	if metav1.IsControlledBy(existingCR, vmSnapshot) {
+		restoredCR := existingCR.DeepCopy()
+		restoredCR.ObjectMeta.Reset()
+
+		// We remove the snapshot name prefix from the CR and add the name of the restored VM
+		restoredCR.Name = vm.Name + strings.Replace(existingCR.Name, vmSnapshotName, "", 1)
+
+		// Make the VirtualMachine the controller of the CR if it already exists, otherwise we will do this later once it has been created.
+		if t.doesTargetVMExist() {
+			restoredCR.OwnerReferences = []metav1.OwnerReference{*metav1.NewControllerRef(vm, kubevirtv1.VirtualMachineGroupVersionKind)}
+		}
```

Comment:
This is confusing to me.
IIUC, if `t.doesTargetVMExist()` is true it means that we're restoring to the same VM, and that basically means that we're going to choose the original CR as this VM's CR. If that's indeed the case, this VM already owns the CR, which must exist already. Why not simply return if that's the case? Or am I missing anything?

I would expect the opposite to happen here - if `t.doesTargetVMExist()` is false, meaning we're performing a clone, we would have to create a new CR and make sure it's owned by this VM. I also don't really understand why we're waiting for `claimInstancetypeControllerRevisionsOwnership()` to do it (which also implicates adding an update action permission to the controller).

Reply:
As above, just covering the corner case where we have to restore the CR for an existing VM whatever reason.

Also, we have to wait and claim ownership for a newly created VirtualMachine as you can't generate an OwnerReference for an object without a UID, we only get the UID after creation.

Reply:
Let me correct myself, as the following statement is wrong:

> ... If that's indeed the case, this VM already owns the CR ...

This isn't true, as the VM could switch to another preference / instancetype.

> Also, we have to wait and claim ownership for a newly created VirtualMachine as you can't generate an OwnerReference for an object without a UID, we only get the UID after creation.

ACK, makes sense :+1: